// @tatumio/tatum/dist/src/blockchain/ada.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cardanoGetUTxos = exports.cardanoGetTransactionsByAccount = exports.cardanoGetAccount = exports.cardanoGetTransaction = exports.cardanoGetBlock = exports.cardanoGetBlockChainInfo = exports.cardanoBroadcast = void 0;
const tatum_1 = require("../connector/tatum");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/CardanoBroadcast" target="_blank">Tatum API documentation</a>
 */
exports.cardanoBroadcast = async (txData) => tatum_1.post(`/v3/cardano/broadcast`, { txData });
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/CardanoGetBlockChainInfo" target="_blank">Tatum API documentation</a>
 */
exports.cardanoGetBlockChainInfo = async () => tatum_1.get(`/v3/cardano/info`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/CardanoGetBlock" target="_blank">Tatum API documentation</a>
 */
exports.cardanoGetBlock = async (hash) => tatum_1.get(`/v3/cardano/block/${hash}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/CardanoGetTransaction" target="_blank">Tatum API documentation</a>
 */
exports.cardanoGetTransaction = async (hash) => tatum_1.get(`/v3/cardano/transaction/${hash}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/CardanoGetAccount" target="_blank">Tatum API documentation</a>
 */
exports.cardanoGetAccount = async (address) => tatum_1.get(`/v3/cardano/account/${address}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/CardanoGetTransactionsByAccount" target="_blank">Tatum API documentation</a>
 */
exports.cardanoGetTransactionsByAccount = async (address, limit, offset) => tatum_1.get(`/v3/cardano/account/${address}/transactions?limit=${limit}?offset=${offset}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/CardanoGetUTxos" target="_blank">Tatum API documentation</a>
 */
exports.cardanoGetUTxos = async (address) => tatum_1.get(`/v3/cardano/${address}/utxos`);
