// @tatumio/tatum/dist/src/blockchain/bcash.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.bcashGetTransaction = exports.bcashGetTxForAccount = exports.bcashGetBlockHash = exports.bcashGetBlock = exports.bcashGetCurrentBlock = exports.bcashBroadcast = void 0;
const tatum_1 = require("../connector/tatum");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/BchBroadcast" target="_blank">Tatum API documentation</a>
 */
exports.bcashBroadcast = async (txData, signatureId) => tatum_1.post(`/v3/bcash/broadcast`, { txData, signatureId });
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/BchGetBlockChainInfo" target="_blank">Tatum API documentation</a>
 */
exports.bcashGetCurrentBlock = async () => tatum_1.get(`/v3/bcash/info`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/BchGetBlock" target="_blank">Tatum API documentation</a>
 */
exports.bcashGetBlock = async (hash) => tatum_1.get(`/v3/bcash/block/${hash}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/BchGetBlockHash" target="_blank">Tatum API documentation</a>
 */
exports.bcashGetBlockHash = async (i) => tatum_1.get(`/v3/bcash/block/hash/${i}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/BchGetTxByAddress" target="_blank">Tatum API documentation</a>
 */
exports.bcashGetTxForAccount = async (address, skip = 0) => tatum_1.get(`/v3/bcash/transaction/address/${address}?skip=${skip}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/BchGetRawTransaction" target="_blank">Tatum API documentation</a>
 */
exports.bcashGetTransaction = async (hash) => tatum_1.get(`/v3/bcash/transaction/${hash}`);
