// @tatumio/tatum/dist/src/blockchain/bitcoin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.btcGetTransaction = exports.btcGetTxForAccount = exports.btcGetUTXO = exports.btcGetBlockHash = exports.btcGetBlock = exports.btcGetBalance = exports.btcGetCurrentBlock = exports.btcBroadcast = void 0;
const tatum_1 = require("../connector/tatum");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/BtcBroadcast" target="_blank">Tatum API documentation</a>
 */
exports.btcBroadcast = async (txData, signatureId) => tatum_1.post(`/v3/bitcoin/broadcast`, { txData, signatureId });
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/BtcGetBlockChainInfo" target="_blank">Tatum API documentation</a>
 */
exports.btcGetCurrentBlock = async () => tatum_1.get('/v3/bitcoin/info');
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/BtcGetBalanceOfAddress" target="_blank">Tatum API documentation</a>
 */
exports.btcGetBalance = async (address) => tatum_1.get(`/v3/bitcoin/address/balance/${address}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/BtcGetBlock" target="_blank">Tatum API documentation</a>
 */
exports.btcGetBlock = async (hash) => tatum_1.get(`/v3/bitcoin/block/${hash}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/BtcGetBlockHash" target="_blank">Tatum API documentation</a>
 */
exports.btcGetBlockHash = async (i) => tatum_1.get(`/v3/bitcoin/block/hash/${i}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/BtcGetUTXO" target="_blank">Tatum API documentation</a>
 */
exports.btcGetUTXO = async (hash, i) => tatum_1.get(`/v3/bitcoin/utxo/${hash}/${i}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/BtcGetTxByAddress" target="_blank">Tatum API documentation</a>
 */
exports.btcGetTxForAccount = async (address, pageSize = 50, offset = 0) => tatum_1.get(`/v3/bitcoin/transaction/address/${address}?pageSize=${pageSize}&offset=${offset}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/BtcGetRawTransaction" target="_blank">Tatum API documentation</a>
 */
exports.btcGetTransaction = async (hash) => tatum_1.get(`/v3/bitcoin/transaction/${hash}`);
