// @tatumio/tatum/dist/src/blockchain/bsc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.bscEstimateGas = exports.bscGetTransaction = exports.bscGetAccountBep20Address = exports.bscGetAccountBalance = exports.bscGetBlock = exports.bscGetCurrentBlock = exports.bscGetTransactionsCount = exports.bscBroadcast = void 0;
const tatum_1 = require("../connector/tatum");
const model_1 = require("../model");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/BscBroadcast" target="_blank">Tatum API documentation</a>
 */
exports.bscBroadcast = async (txData, signatureId) => tatum_1.post(`/v3/bsc/broadcast`, { txData, signatureId });
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/BscGetTransactionCount" target="_blank">Tatum API documentation</a>
 */
exports.bscGetTransactionsCount = async (address) => tatum_1.get(`/v3/bsc/transaction/count/${address}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/BscGetCurrentBlock" target="_blank">Tatum API documentation</a>
 */
exports.bscGetCurrentBlock = async () => tatum_1.get(`/v3/bsc/block/current`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/BscGetBlock" target="_blank">Tatum API documentation</a>
 */
exports.bscGetBlock = async (hash) => tatum_1.get(`/v3/bsc/block/${hash}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/BscGetBalance" target="_blank">Tatum API documentation</a>
 */
exports.bscGetAccountBalance = async (address) => tatum_1.get(`/v3/bsc/account/balance/${address}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/BscBep20GetBalance" target="_blank">Tatum API documentation</a>
 */
exports.bscGetAccountBep20Address = async (address, contractAddress) => tatum_1.get(`/v3/bsc/account/balance/bep20/${address}?contractAddress=${contractAddress}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/BscGetTransaction" target="_blank">Tatum API documentation</a>
 */
exports.bscGetTransaction = async (hash) => tatum_1.get(`/v3/bsc/transaction/${hash}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/BscEstimateGas" target="_blank">Tatum API documentation</a>
 */
exports.bscEstimateGas = (body) => tatum_1.post('/v3/bsc/gas', body, model_1.EstimateGasEth);
