// @tatumio/tatum/dist/src/blockchain/celo.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.celoGetTransaction = exports.celoGetAccountBalance = exports.celoGetBlock = exports.celoGetCurrentBlock = exports.celoGetTransactionsCount = exports.celoBroadcast = void 0;
const tatum_1 = require("../connector/tatum");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/CeloBroadcast" target="_blank">Tatum API documentation</a>
 */
exports.celoBroadcast = async (txData, signatureId) => tatum_1.post(`/v3/celo/broadcast`, {
    txData,
    signatureId
});
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/CeloGetTransactionCount" target="_blank">Tatum API documentation</a>
 */
exports.celoGetTransactionsCount = async (address) => tatum_1.get(`/v3/celo/transaction/count/${address}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/CeloGetCurrentBlock" target="_blank">Tatum API documentation</a>
 */
exports.celoGetCurrentBlock = async () => tatum_1.get(`/v3/celo/block/current`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/CeloGetBlock" target="_blank">Tatum API documentation</a>
 */
exports.celoGetBlock = async (hash) => tatum_1.get(`/v3/celo/block/${hash}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/CeloGetBalance" target="_blank">Tatum API documentation</a>
 */
exports.celoGetAccountBalance = async (address) => {
    const { data } = await tatum_1.get(`/v3/celo/account/balance/${address}`);
    return { celo: new bignumber_js_1.default(data.celo), cUsd: new bignumber_js_1.default(data.cUsd) };
};
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/CeloGetTransaction" target="_blank">Tatum API documentation</a>
 */
exports.celoGetTransaction = async (hash) => tatum_1.get(`/v3/celo/transaction/${hash}`);
