// @tatumio/tatum/dist/src/blockchain/dogecoin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dogeGetTransaction = exports.dogeGetUTXO = exports.dogeGetBlockHash = exports.dogeGetBlock = exports.dogeGetCurrentBlock = exports.dogeBroadcast = void 0;
const tatum_1 = require("../connector/tatum");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/DogeBroadcast" target="_blank">Tatum API documentation</a>
 */
exports.dogeBroadcast = async (txData, signatureId) => tatum_1.post(`/v3/dogecoin/broadcast`, { txData, signatureId });
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/DogeGetBlockChainInfo" target="_blank">Tatum API documentation</a>
 */
exports.dogeGetCurrentBlock = async () => tatum_1.get(`/v3/dogecoin/info`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/DogeGetBlock" target="_blank">Tatum API documentation</a>
 */
exports.dogeGetBlock = async (hash) => tatum_1.get(`/v3/dogecoin/block/${hash}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/DogeGetBlockHash" target="_blank">Tatum API documentation</a>
 */
exports.dogeGetBlockHash = async (i) => tatum_1.get(`/v3/dogecoin/block/hash/${i}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/DogeGetUTXO" target="_blank">Tatum API documentation</a>
 */
exports.dogeGetUTXO = async (hash, i) => tatum_1.get(`/v3/dogecoin/utxo/${hash}/${i}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/DogeGetRawTransaction" target="_blank">Tatum API documentation</a>
 */
exports.dogeGetTransaction = async (hash) => tatum_1.get(`/v3/dogecoin/transaction/${hash}`);
