// @tatumio/tatum/dist/src/blockchain/ethereum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ethEstimateGas = exports.ethGetAccountTransactions = exports.ethGetTransaction = exports.ethGetAccountErc20Address = exports.ethGetAccountBalance = exports.ethGetBlock = exports.ethGetCurrentBlock = exports.ethGetTransactionsCount = exports.ethBroadcast = void 0;
const tatum_1 = require("../connector/tatum");
const model_1 = require("../model");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/EthBroadcast" target="_blank">Tatum API documentation</a>
 */
exports.ethBroadcast = async (txData, signatureId) => tatum_1.post(`/v3/ethereum/broadcast`, { txData, signatureId });
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/EthGetTransactionCount" target="_blank">Tatum API documentation</a>
 */
exports.ethGetTransactionsCount = async (address) => tatum_1.get(`/v3/ethereum/transaction/count/${address}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/EthGetCurrentBlock" target="_blank">Tatum API documentation</a>
 */
exports.ethGetCurrentBlock = async () => tatum_1.get(`/v3/ethereum/block/current`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/EthGetBlock" target="_blank">Tatum API documentation</a>
 */
exports.ethGetBlock = async (hash) => tatum_1.get(`/v3/ethereum/block/${hash}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/EthGetBalance" target="_blank">Tatum API documentation</a>
 */
exports.ethGetAccountBalance = async (address) => tatum_1.get(`/v3/ethereum/account/balance/${address}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/EthErc20GetBalance" target="_blank">Tatum API documentation</a>
 */
exports.ethGetAccountErc20Address = async (address, contractAddress) => tatum_1.get(`/v3/ethereum/account/balance/erc20/${address}?contractAddress=${contractAddress}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/EthGetTransaction" target="_blank">Tatum API documentation</a>
 */
exports.ethGetTransaction = async (hash) => tatum_1.get(`/v3/ethereum/transaction/${hash}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/EthGetTransactionByAddress" target="_blank">Tatum API documentation</a>
 */
exports.ethGetAccountTransactions = async (address, pageSize = 50, offset = 0) => tatum_1.get(`/v3/ethereum/account/transaction/${address}?pageSize=${pageSize}&offset=${offset}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/EthEstimateGas" target="_blank">Tatum API documentation</a>
 */
exports.ethEstimateGas = (body) => tatum_1.post('/v3/ethereum/gas', body, model_1.EstimateGasEth);
