// @tatumio/tatum/dist/src/blockchain/litecoin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ltcGetTransaction = exports.ltcGetTxForAccount = exports.ltcGetUTXO = exports.ltcGetBlockHash = exports.ltcGetBlock = exports.ltcGetCurrentBlock = exports.ltcGetBalance = exports.ltcBroadcast = void 0;
const tatum_1 = require("../connector/tatum");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/LtcBroadcast" target="_blank">Tatum API documentation</a>
 */
exports.ltcBroadcast = async (txData, signatureId) => tatum_1.post(`/v3/litecoin/broadcast`, { txData, signatureId });
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/LtcGetBalanceOfAddress" target="_blank">Tatum API documentation</a>
 */
exports.ltcGetBalance = async (address) => tatum_1.get(`/v3/litecoin/address/balance/${address}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/LtcGetBlockChainInfo" target="_blank">Tatum API documentation</a>
 */
exports.ltcGetCurrentBlock = async () => tatum_1.get(`/v3/litecoin/info`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/LtcGetBlock" target="_blank">Tatum API documentation</a>
 */
exports.ltcGetBlock = async (hash) => tatum_1.get(`/v3/litecoin/block/${hash}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/LtcGetBlockHash" target="_blank">Tatum API documentation</a>
 */
exports.ltcGetBlockHash = async (i) => tatum_1.get(`/v3/litecoin/block/hash/${i}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/LtcGetUTXO" target="_blank">Tatum API documentation</a>
 */
exports.ltcGetUTXO = async (hash, i) => tatum_1.get(`/v3/litecoin/utxo/${hash}/${i}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/LtcGetTxByAddress" target="_blank">Tatum API documentation</a>
 */
exports.ltcGetTxForAccount = async (address, pageSize = 50, offset = 0) => tatum_1.get(`/v3/litecoin/transaction/address/${address}?pageSize=${pageSize}&offset=${offset}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/LtcGetRawTransaction" target="_blank">Tatum API documentation</a>
 */
exports.ltcGetTransaction = async (hash) => tatum_1.get(`/v3/litecoin/transaction/${hash}`);
