// @tatumio/tatum/dist/src/blockchain/quorum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.quorumUnlockAccount = exports.quorumCreateAccount = exports.quorumGetTransactionReceipt = exports.quorumGetTransaction = exports.quorumGetBlock = exports.quorumGetCurrentBlock = void 0;
const tatum_1 = require("../connector/tatum");
const model_1 = require("../model");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/QuorumGetCurrentBlock" target="_blank">Tatum API documentation</a>
 */
exports.quorumGetCurrentBlock = async () => tatum_1.get(`/v3/quorum/block/current`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/QuorumGetBlock" target="_blank">Tatum API documentation</a>
 */
exports.quorumGetBlock = async (hash) => tatum_1.get(`/v3/quorum/block/${hash}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/QuorumGetTransaction" target="_blank">Tatum API documentation</a>
 */
exports.quorumGetTransaction = async (hash) => tatum_1.get(`/v3/quorum/transaction/${hash}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/QuorumGetTransactionReceipt" target="_blank">Tatum API documentation</a>
 */
exports.quorumGetTransactionReceipt = async (hash) => tatum_1.get(`/v3/quorum/transaction/${hash}/receipt`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/QuorumGenerateAccount" target="_blank">Tatum API documentation</a>
 */
exports.quorumCreateAccount = async (body) => tatum_1.post(`/v3/quorum/account`, body, model_1.AccountPassword);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/QuorumUnlockAccount" target="_blank">Tatum API documentation</a>
 */
exports.quorumUnlockAccount = async (address, body) => tatum_1.post(`/v3/quorum/account/${address}/unlock`, body, model_1.AccountPassword);
