// @tatumio/tatum/dist/src/blockchain/scrypta.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.scryptaGetTransaction = exports.scryptaGetUnspentForAccount = exports.scryptaGetTxForAccount = exports.scryptaGetUTXO = exports.scryptaGetBlockHash = exports.scryptaGetBlock = exports.scryptaGetCurrentBlock = exports.scryptaBroadcast = void 0;
const tatum_1 = require("../connector/tatum");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/BroadcastsignedScryptatransaction" target="_blank">Tatum API documentation</a>
 */
exports.scryptaBroadcast = async (txData, signatureId) => tatum_1.post(`/v3/scrypta/broadcast`, { txData, signatureId });
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/ScryptaGetBlockChainInfo" target="_blank">Tatum API documentation</a>
 */
exports.scryptaGetCurrentBlock = async () => tatum_1.get(`/v3/scrypta/info`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/ScryptaGetBlock" target="_blank">Tatum API documentation</a>
 */
exports.scryptaGetBlock = async (hash) => tatum_1.get(`/v3/scrypta/block/${hash}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/ScryptaGetBlockHash" target="_blank">Tatum API documentation</a>
 */
exports.scryptaGetBlockHash = async (i) => tatum_1.get(`/v3/scrypta/block/hash/${i}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/ScryptaGetUTXO" target="_blank">Tatum API documentation</a>
 */
exports.scryptaGetUTXO = async (hash, i) => tatum_1.get(`/v3/scrypta/utxo/${hash}/${i}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/ScryptaGetTxByAddress" target="_blank">Tatum API documentation</a>
 */
exports.scryptaGetTxForAccount = async (address, pageSize = 50, offset = 0) => tatum_1.get(`/v3/scrypta/transaction/address/${address}?pageSize=${pageSize}&offset=${offset}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/ScryptaGetTxByAddress" target="_blank">Tatum API documentation</a>
 */
exports.scryptaGetUnspentForAccount = async (address, pageSize = 50, offset = 0) => tatum_1.get(`/v3/scrypta/utxo/${address}?pageSize=${pageSize}&offset=${offset}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/ScryptaGetRawTransaction" target="_blank">Tatum API documentation</a>
 */
exports.scryptaGetTransaction = async (hash) => tatum_1.get(`/v3/scrypta/transaction/${hash}`);
