// @tatumio/tatum/dist/src/blockchain/tron.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tronGetAccount = exports.tronGetTransaction = exports.tronGetTrc10Detail = exports.tronGetBlock = exports.tronGetCurrentBlock = exports.tronBroadcast = void 0;
const tatum_1 = require("../connector/tatum");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/TronBroadcast" target="_blank">Tatum API documentation</a>
 */
exports.tronBroadcast = async (txData, signatureId) => tatum_1.post(`/v3/tron/broadcast`, { txData, signatureId });
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/TronGetCurrentBlock" target="_blank">Tatum API documentation</a>
 */
exports.tronGetCurrentBlock = async () => tatum_1.get(`/v3/tron/current`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/TronGetBlock" target="_blank">Tatum API documentation</a>
 */
exports.tronGetBlock = async (hash) => tatum_1.get(`/v3/tron/block/${hash}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/TronTrc10Detail" target="_blank">Tatum API documentation</a>
 */
exports.tronGetTrc10Detail = async (id) => tatum_1.get(`/v3/tron/trc10/detail/${id}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/TronGetTransaction" target="_blank">Tatum API documentation</a>
 */
exports.tronGetTransaction = async (hash) => tatum_1.get(`/v3/tron/transaction/${hash}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/TronGetAccount" target="_blank">Tatum API documentation</a>
 */
exports.tronGetAccount = async (address) => tatum_1.get(`/v3/tron/account/${address}`);
