// @tatumio/tatum/dist/src/blockchain/xlm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.xlmGetAccountTransactions = exports.xlmGetTransaction = exports.xlmGetLedgerTx = exports.xlmGetLedger = exports.xlmGetFee = exports.xlmGetCurrentLedger = exports.xlmBroadcast = exports.xlmGetAccountInfo = void 0;
const tatum_1 = require("../connector/tatum");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XlmGetAccountInfo" target="_blank">Tatum API documentation</a>
 */
exports.xlmGetAccountInfo = async (account) => tatum_1.get(`/v3/xlm/account/${account}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XlmBroadcast" target="_blank">Tatum API documentation</a>
 */
exports.xlmBroadcast = async (txData, signatureId) => tatum_1.post(`/v3/xlm/broadcast`, { txData, signatureId });
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XlmGetLastClosedLedger" target="_blank">Tatum API documentation</a>
 */
exports.xlmGetCurrentLedger = async () => tatum_1.get(`/v3/xlm/info`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XlmGetFee" target="_blank">Tatum API documentation</a>
 */
exports.xlmGetFee = async () => tatum_1.get(`/v3/xlm/fee`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XlmGetLedger" target="_blank">Tatum API documentation</a>
 */
exports.xlmGetLedger = async (i) => tatum_1.get(`/v3/xlm/ledger/${i}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XlmGetLedgerTx" target="_blank">Tatum API documentation</a>
 */
exports.xlmGetLedgerTx = async (i) => tatum_1.get(`/v3/xlm/ledger/${i}/transaction`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XlmGetTransaction" target="_blank">Tatum API documentation</a>
 */
exports.xlmGetTransaction = async (hash) => tatum_1.get(`/v3/xlm/transaction/${hash}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XlmGetAccountTx" target="_blank">Tatum API documentation</a>
 */
exports.xlmGetAccountTransactions = async (address) => tatum_1.get(`/v3/xlm/account/tx/${address}`);
