// @tatumio/tatum/dist/src/blockchain/xrp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.xrpGetAccountTransactions = exports.xrpGetTransaction = exports.xrpGetAccountBalance = exports.xrpGetLedger = exports.xrpGetCurrentLedger = exports.xrpBroadcast = exports.xrpGetAccountInfo = exports.xrpGetFee = void 0;
const tatum_1 = require("../connector/tatum");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XrpGetFee" target="_blank">Tatum API documentation</a>
 */
exports.xrpGetFee = async () => tatum_1.get(`/v3/xrp/fee`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XrpGetAccountInfo" target="_blank">Tatum API documentation</a>
 */
exports.xrpGetAccountInfo = async (account) => tatum_1.get(`/v3/xrp/account/${account}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XrpBroadcast" target="_blank">Tatum API documentation</a>
 */
exports.xrpBroadcast = async (txData, signatureId) => tatum_1.post(`/v3/xrp/broadcast`, { txData, signatureId });
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XrpGetLastClosedLedger" target="_blank">Tatum API documentation</a>
 */
exports.xrpGetCurrentLedger = async () => tatum_1.get(`/v3/xrp/info`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XrpGetLedger" target="_blank">Tatum API documentation</a>
 */
exports.xrpGetLedger = async (i) => tatum_1.get(`/v3/xrp/ledger/${i}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XrpGetAccountBalance" target="_blank">Tatum API documentation</a>
 */
exports.xrpGetAccountBalance = async (address) => tatum_1.get(`/v3/xrp/account/${address}/balance`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XrpGetTransaction" target="_blank">Tatum API documentation</a>
 */
exports.xrpGetTransaction = async (hash) => tatum_1.get(`/v3/xrp/transaction/${hash}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XrpGetAccountTx" target="_blank">Tatum API documentation</a>
 */
exports.xrpGetAccountTransactions = async (address, min, marker) => tatum_1.get(`/v3/xrp/account/tx/${address}?min=${min}&marker=${marker}`);
