// @tatumio/tatum/dist/src/nft/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transferNFT = exports.burnNFT = exports.mintMultipleNFTWithUri = exports.mintNFTWithUri = exports.deployNFT = exports.getNFTMetadataURI = exports.getNFTsByAddress = void 0;
const tatum_1 = require("../connector/tatum");
const model_1 = require("../model");
const transaction_1 = require("../transaction");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/NftGetBalanceErc721" target="_blank">Tatum API documentation</a>
 */
exports.getNFTsByAddress = async (chain, contractAddress, address) => tatum_1.get(`/v3/nft/balance/${chain}/${contractAddress}/${address}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/NftGetMetadataErc721" target="_blank">Tatum API documentation</a>
 */
exports.getNFTMetadataURI = async (chain, contractAddress, tokenId) => tatum_1.get(`/v3/nft/metadata/${chain}/${contractAddress}/${tokenId}`);
exports.deployNFT = async (testnet, body, provider) => {
    switch (body.chain) {
        case model_1.Currency.CELO:
            return transaction_1.sendCeloDeployErc721Transaction(testnet, body, provider);
        case model_1.Currency.ETH:
            return transaction_1.sendDeployErc721Transaction(body, provider);
        case model_1.Currency.BSC:
            return transaction_1.sendDeployBep721Transaction(body, provider);
    }
};
exports.mintNFTWithUri = async (testnet, body, provider) => {
    switch (body.chain) {
        case model_1.Currency.CELO:
            return transaction_1.sendCeloMinErc721Transaction(testnet, body, provider);
        case model_1.Currency.ETH:
            return transaction_1.sendMintErc721Transaction(body, provider);
        case model_1.Currency.BSC:
            return transaction_1.sendMintBep721Transaction(body, provider);
    }
};
exports.mintMultipleNFTWithUri = async (testnet, body, provider) => {
    switch (body.chain) {
        case model_1.Currency.CELO:
            return transaction_1.sendCeloMintMultipleErc721Transaction(testnet, body, provider);
        case model_1.Currency.ETH:
            return transaction_1.sendMintMultipleErc721Transaction(body, provider);
        case model_1.Currency.BSC:
            return transaction_1.sendMintMultipleBep721Transaction(body, provider);
    }
};
exports.burnNFT = async (testnet, body, provider) => {
    switch (body.chain) {
        case model_1.Currency.CELO:
            return transaction_1.sendCeloBurnErc721Transaction(testnet, body, provider);
        case model_1.Currency.ETH:
            return transaction_1.sendBurnErc721Transaction(body, provider);
        case model_1.Currency.BSC:
            return transaction_1.sendBurnBep721Transaction(body, provider);
    }
};
exports.transferNFT = async (testnet, body, provider) => {
    switch (body.chain) {
        case model_1.Currency.CELO:
            return transaction_1.sendCeloTransferErc721Transaction(testnet, body, provider);
        case model_1.Currency.ETH:
            return transaction_1.sendErc721Transaction(body, provider);
        case model_1.Currency.BSC:
            return transaction_1.sendBep721Transaction(body, provider);
    }
};
