// @tatumio/tatum/dist/src/nft/nft.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const model_1 = require("../model");
const index_1 = require("./index");
describe('NFT tests', () => {
    jest.setTimeout(19999);
    describe('NFT CELO transactions', () => {
        // ERC-721
        it('should test valid deploy 721 transaction', async () => {
            const body = new model_1.CeloDeployErc721();
            body.fromPrivateKey = '0x4874827a55d87f2309c55b835af509e3427aa4d52321eeb49a2b93b5c0f8edfb';
            body.name = 'Tatum';
            body.symbol = 'TTM';
            body.feeCurrency = model_1.Currency.CUSD;
            body.chain = model_1.Currency.CELO;
            expect(await index_1.deployNFT(true, body, 'https://alfajores-forno.celo-testnet.org')).toBeDefined();
        });
        it('should test valid mint 721 transaction', async () => {
            const body = new model_1.CeloMintErc721();
            body.fromPrivateKey = '0x4874827a55d87f2309c55b835af509e3427aa4d52321eeb49a2b93b5c0f8edfb';
            body.to = '0x8cb76aed9c5e336ef961265c6079c14e9cd3d2ea';
            body.contractAddress = '0x3e1a302DA9345ae6f8188607C017d342A4CCf22e';
            body.tokenId = '3';
            body.url = 'https://google.com';
            body.feeCurrency = model_1.Currency.CUSD;
            body.chain = model_1.Currency.CELO;
            expect(await index_1.mintNFTWithUri(true, body, 'https://alfajores-forno.celo-testnet.org')).toBeDefined();
        });
        it('should test valid mint multiple 721 transaction', async () => {
            const body = new model_1.CeloMintMultipleErc721();
            body.fromPrivateKey = '0x4874827a55d87f2309c55b835af509e3427aa4d52321eeb49a2b93b5c0f8edfb';
            body.to = ['0x8cb76aed9c5e336ef961265c6079c14e9cd3d2ea', '0x8cb76aed9c5e336ef961265c6079c14e9cd3d2ea'];
            body.contractAddress = '0x3e1a302DA9345ae6f8188607C017d342A4CCf22e';
            body.tokenId = ['4', '5'];
            body.url = ['https://google.com', 'https://google.com'];
            body.feeCurrency = model_1.Currency.CUSD;
            body.chain = model_1.Currency.CELO;
            expect(await index_1.mintMultipleNFTWithUri(true, body, 'https://alfajores-forno.celo-testnet.org')).toBeDefined();
        });
        it('should test valid burn 721 transaction', async () => {
            const body = new model_1.CeloBurnErc721();
            body.fromPrivateKey = '0x4874827a55d87f2309c55b835af509e3427aa4d52321eeb49a2b93b5c0f8edfb';
            body.contractAddress = '0x3e1a302DA9345ae6f8188607C017d342A4CCf22e';
            body.tokenId = '3';
            body.feeCurrency = model_1.Currency.CUSD;
            body.chain = model_1.Currency.CELO;
            expect(await index_1.burnNFT(true, body, 'https://alfajores-forno.celo-testnet.org')).toBeDefined();
        });
        it('should test valid transfer 721 transaction', async () => {
            const body = new model_1.CeloTransferErc721();
            body.fromPrivateKey = '0x4874827a55d87f2309c55b835af509e3427aa4d52321eeb49a2b93b5c0f8edfb';
            body.contractAddress = '0x3e1a302DA9345ae6f8188607C017d342A4CCf22e';
            body.to = '0x10168acf3231ccc7b16ba53f17dd4d8bdecf4e1a';
            body.tokenId = '5';
            body.feeCurrency = model_1.Currency.CUSD;
            body.chain = model_1.Currency.CELO;
            expect(await index_1.transferNFT(true, body, 'https://alfajores-forno.celo-testnet.org')).toBeDefined();
        });
    });
    describe('NFT ETH transactions', () => {
        it('should test eth 721 deploy transaction', async () => {
            const deployErc721Token = await index_1.deployNFT(true, {
                symbol: '1oido3id3',
                chain: model_1.Currency.ETH,
                fromPrivateKey: '0x1a4344e55c562db08700dd32e52e62e7c40b1ef5e27c6ddd969de9891a899b29',
                name: '2123kd',
            });
            expect(deployErc721Token).not.toBeNull();
            console.log(deployErc721Token);
        });
        it('should test eth 721 mint transaction', async () => {
            const tokenId = new Date().getTime().toString();
            const mintedToken = await index_1.mintNFTWithUri(true, {
                to: '0x687422eEA2cB73B5d3e242bA5456b782919AFc85',
                chain: model_1.Currency.ETH,
                tokenId,
                url: 'https://www.seznam.cz',
                fromPrivateKey: '0x1a4344e55c562db08700dd32e52e62e7c40b1ef5e27c6ddd969de9891a899b29',
                contractAddress: '0xAB12C6c926Cc3c9547aad71d6082fA724152A442',
                fee: {
                    gasLimit: '500000',
                    gasPrice: '110'
                }
            });
            console.log(tokenId);
            expect(mintedToken).not.toBeNull();
        });
        it('should test eth 721 mint multiple transaction', async () => {
            const firstTokenId = new Date().getTime();
            const secondTokenId = firstTokenId + 1;
            const mintedTokens = await index_1.mintMultipleNFTWithUri(true, {
                to: ['0x811dfbff13adfbc3cf653dcc373c03616d3471c9', '0x811dfbff13adfbc3cf653dcc373c03616d3471c9'],
                chain: model_1.Currency.ETH,
                tokenId: [firstTokenId.toString(), secondTokenId.toString()],
                url: ['https://www.seznam.cz', 'https://www.seznam.cz'],
                fromPrivateKey: '0x1a4344e55c562db08700dd32e52e62e7c40b1ef5e27c6ddd969de9891a899b29',
                contractAddress: '0xab12c6c926cc3c9547aad71d6082fa724152a442',
                fee: {
                    gasLimit: '500000',
                    gasPrice: '100'
                }
            });
            console.log(mintedTokens);
            expect(mintedTokens).not.toBeNull();
        });
        it('should test eth 721 burn transaction', async () => {
            const burnErc721Token = await index_1.burnNFT(true, {
                tokenId: '1615884747446',
                chain: model_1.Currency.ETH,
                fromPrivateKey: '0x1a4344e55c562db08700dd32e52e62e7c40b1ef5e27c6ddd969de9891a899b29',
                contractAddress: '0xab12c6c926cc3c9547aad71d6082fa724152a442',
                fee: {
                    gasLimit: '5000000',
                    gasPrice: '1100'
                },
            });
            expect(burnErc721Token).not.toBeNull();
        });
        it('should test eth 721 send transaction', async () => {
            const sendErc721Token = await index_1.transferNFT(true, {
                to: '0x811dfbff13adfbc3cf653dcc373c03616d3471c9',
                chain: model_1.Currency.ETH,
                tokenId: '1615884907854',
                fromPrivateKey: '0x1a4344e55c562db08700dd32e52e62e7c40b1ef5e27c6ddd969de9891a899b29',
                contractAddress: '0xab12c6c926cc3c9547aad71d6082fa724152a442',
                fee: {
                    gasLimit: '5000000',
                    gasPrice: '100'
                },
            });
            expect(sendErc721Token).not.toBeNull();
        });
    });
});
