// @tatumio/tatum/dist/src/model/request/Currency.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BSC_BASED_CURRENCIES = exports.ETH_BASED_CURRENCIES = exports.BEP20_CURRENCIES = exports.ERC20_CURRENCIES = exports.Currency = void 0;
var Currency;
(function (Currency) {
    Currency["BTC"] = "BTC";
    Currency["BCH"] = "BCH";
    Currency["LTC"] = "LTC";
    Currency["CELO"] = "CELO";
    Currency["CUSD"] = "CUSD";
    Currency["ETH"] = "ETH";
    Currency["FABRIC"] = "FABRIC";
    Currency["QUORUM"] = "QUORUM";
    Currency["XRP"] = "XRP";
    Currency["XLM"] = "XLM";
    Currency["DOGE"] = "DOGE";
    Currency["VET"] = "VET";
    Currency["NEO"] = "NEO";
    Currency["BNB"] = "BNB";
    Currency["BSC"] = "BSC";
    Currency["BETH"] = "BETH";
    Currency["BBTC"] = "BBTC";
    Currency["BADA"] = "BADA";
    Currency["WBNB"] = "WBNB";
    Currency["BDOT"] = "BDOT";
    Currency["BXRP"] = "BXRP";
    Currency["BLTC"] = "BLTC";
    Currency["BBCH"] = "BBCH";
    Currency["USDT"] = "USDT";
    Currency["USDT_TRON"] = "USDT_TRON";
    Currency["TRON"] = "TRON";
    Currency["LEO"] = "LEO";
    Currency["LINK"] = "LINK";
    Currency["WBTC"] = "WBTC";
    Currency["UNI"] = "UNI";
    Currency["FREE"] = "FREE";
    Currency["MKR"] = "MKR";
    Currency["USDC"] = "USDC";
    Currency["BAT"] = "BAT";
    Currency["TUSD"] = "TUSD";
    Currency["PAX"] = "PAX";
    Currency["PLTC"] = "PLTC";
    Currency["XCON"] = "XCON";
    Currency["MMY"] = "MMY";
    Currency["PAXG"] = "PAXG";
    Currency["LYRA"] = "LYRA";
    Currency["ADA"] = "ADA";
})(Currency = exports.Currency || (exports.Currency = {}));
exports.ERC20_CURRENCIES = [
    Currency.USDT.toString(),
    Currency.LEO.toString(),
    Currency.LINK.toString(),
    Currency.UNI.toString(),
    Currency.FREE.toString(),
    Currency.MKR.toString(),
    Currency.USDC.toString(),
    Currency.BAT.toString(),
    Currency.WBTC.toString(),
    Currency.TUSD.toString(),
    Currency.PAX.toString(),
    Currency.PAXG.toString(),
    Currency.PLTC.toString(),
    Currency.XCON.toString(),
    Currency.MMY.toString(),
];
exports.BEP20_CURRENCIES = [
    Currency.BETH.toString(),
    Currency.BBTC.toString(),
    Currency.BADA.toString(),
    Currency.WBNB.toString(),
    Currency.BDOT.toString(),
    Currency.BXRP.toString(),
    Currency.BLTC.toString(),
    Currency.BBCH.toString(),
];
exports.ETH_BASED_CURRENCIES = [
    Currency.ETH.toString(),
    ...exports.ERC20_CURRENCIES,
];
exports.BSC_BASED_CURRENCIES = [
    Currency.BSC.toString(),
    ...exports.BEP20_CURRENCIES,
];
