// @tatumio/tatum/dist/src/ledger/account.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAccountBalance = exports.getAllAccounts = exports.getAccountsByCustomerId = exports.unfreezeAccount = exports.freezeAccount = exports.deactivateAccount = exports.activateAccount = exports.deleteBlockedAmountForAccount = exports.deleteBlockedAmountWithTransaction = exports.deleteBlockedAmount = exports.blockAmount = exports.getBlockedAmountsByAccountId = exports.createAccounts = exports.updateAccount = exports.createAccount = exports.getAccountById = void 0;
const tatum_1 = require("../connector/tatum");
const model_1 = require("../model");
const BlockageTransaction_1 = require("../model/request/BlockageTransaction");
const CreateAccountsBatch_1 = require("../model/request/CreateAccountsBatch");
const UpdateAccount_1 = require("../model/request/UpdateAccount");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/getAccountByAccountId" target="_blank">Tatum API documentation</a>
 */
exports.getAccountById = async (id) => tatum_1.get(`/v3/ledger/account/${id}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/createAccount" target="_blank">Tatum API documentation</a>
 */
exports.createAccount = async (account) => tatum_1.post('/v3/ledger/account', account, model_1.CreateAccount);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/updateAccountByAccountId" target="_blank">Tatum API documentation</a>
 */
exports.updateAccount = async (id, account) => tatum_1.put(`/v3/ledger/account/${id}`, account, UpdateAccount_1.UpdateAccount);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/createAccountBatch" target="_blank">Tatum API documentation</a>
 */
exports.createAccounts = async (accounts) => tatum_1.post('/v3/ledger/account/batch', accounts, CreateAccountsBatch_1.CreateAccountsBatch);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/getBlockAmount" target="_blank">Tatum API documentation</a>
 */
exports.getBlockedAmountsByAccountId = async (id, pageSize = 50, offset = 0) => tatum_1.get(`/v3/ledger/account/block/${id}?pageSize=${pageSize}&offset=${offset}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/blockAmount" target="_blank">Tatum API documentation</a>
 */
exports.blockAmount = async (id, block) => tatum_1.post(`/v3/ledger/account/block/${id}`, block, model_1.BlockAmount);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/deleteBlockAmount" target="_blank">Tatum API documentation</a>
 */
exports.deleteBlockedAmount = async (id) => tatum_1.httpDelete(`/v3/ledger/account/block/${id}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/unblockAmountWithTransaction" target="_blank">Tatum API documentation</a>
 */
exports.deleteBlockedAmountWithTransaction = async (id, txData) => tatum_1.put(`/v3/ledger/account/block/${id}`, txData, BlockageTransaction_1.BlockageTransaction);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/deleteAllBlockAmount" target="_blank">Tatum API documentation</a>
 */
exports.deleteBlockedAmountForAccount = async (id) => tatum_1.httpDelete(`/v3/ledger/account/block/account/${id}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/activateAccount" target="_blank">Tatum API documentation</a>
 */
exports.activateAccount = async (id) => tatum_1.put(`v3/ledger/account/${id}/activate`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/deactivateAccount" target="_blank">Tatum API documentation</a>
 */
exports.deactivateAccount = async (id) => tatum_1.put(`/v3/ledger/account/${id}/deactivate`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/freezeAccount" target="_blank">Tatum API documentation</a>
 */
exports.freezeAccount = async (id) => tatum_1.put(`/v3/ledger/account/${id}/freeze`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/unfreezeAccount" target="_blank">Tatum API documentation</a>
 */
exports.unfreezeAccount = async (id) => tatum_1.put(`/v3/ledger/account/${id}/unfreeze`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/getAccountsByCustomerId" target="_blank">Tatum API documentation</a>
 */
exports.getAccountsByCustomerId = async (id, pageSize = 50, offset = 0) => tatum_1.get(`/v3/ledger/account/customer/${id}?pageSize=${pageSize}&offset=${offset}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/getAllAccounts" target="_blank">Tatum API documentation</a>
 */
exports.getAllAccounts = async (pageSize = 50, offset = 0) => tatum_1.get(`/v3/ledger/account?pageSize=${pageSize}&offset=${offset}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/getAccountBalance" target="_blank">Tatum API documentation</a>
 */
exports.getAccountBalance = async (id) => tatum_1.get(`/v3/ledger/account/${id}/balance`);
