// @tatumio/tatum/dist/src/ledger/customer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.disableCustomer = exports.enableCustomer = exports.deactivateCustomer = exports.activateCustomer = exports.updateCustomer = exports.getAllCustomers = exports.getCustomer = void 0;
const tatum_1 = require("../connector/tatum");
const CustomerUpdate_1 = require("../model/request/CustomerUpdate");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/getCustomerByExternalId" target="_blank">Tatum API documentation</a>
 */
exports.getCustomer = async (id) => tatum_1.get(`/v3/ledger/customer/${id}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/findAllCustomers" target="_blank">Tatum API documentation</a>
 */
exports.getAllCustomers = async (pageSize = 50, offset = 0) => tatum_1.get(`/v3/ledger/customer?pageSize=${pageSize}&offset=${offset}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/updateCustomer" target="_blank">Tatum API documentation</a>
 */
exports.updateCustomer = async (id, data) => tatum_1.put(`/v3/ledger/customer/${id}`, data, CustomerUpdate_1.CustomerUpdate);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/activateAccount" target="_blank">Tatum API documentation</a>
 */
exports.activateCustomer = async (id) => tatum_1.put(`/v3/ledger/customer/${id}/activate`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/deactivateCustomer" target="_blank">Tatum API documentation</a>
 */
exports.deactivateCustomer = async (id) => tatum_1.put(`/v3/ledger/customer/${id}/deactivate`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/enableCustomer" target="_blank">Tatum API documentation</a>
 */
exports.enableCustomer = async (id) => tatum_1.put(`/v3/ledger/customer/${id}/enable`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/disableCustomer" target="_blank">Tatum API documentation</a>
 */
exports.disableCustomer = async (id) => tatum_1.put(`/v3/ledger/customer/${id}/disable`);
