// @tatumio/tatum/dist/src/ledger/transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.countTransactionsByLedger = exports.countTransactionsByCustomer = exports.countTransactionsByAccount = exports.getTransactionsByLedger = exports.getTransactionsByCustomer = exports.getTransactionsByAccount = exports.storeTransaction = exports.getTransactionsByReference = void 0;
const tatum_1 = require("../connector/tatum");
const model_1 = require("../model");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/getTransactionsByReference" target="_blank">Tatum API documentation</a>
 */
exports.getTransactionsByReference = async (reference) => tatum_1.get(`/v3/ledger/transaction/reference/${reference}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/sendTransaction" target="_blank">Tatum API documentation</a>
 */
exports.storeTransaction = async (transaction) => tatum_1.post(`/v3/ledger/transaction`, transaction, model_1.CreateTransaction);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/getTransactionsByAccountId" target="_blank">Tatum API documentation</a>
 */
exports.getTransactionsByAccount = async (filter, pageSize = 50, offset = 0) => tatum_1.post(`/v3/ledger/transaction/account?pageSize=${pageSize}&offset=${offset}`, filter, model_1.TransactionFilter);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/getTransactionsByCustomerId" target="_blank">Tatum API documentation</a>
 */
exports.getTransactionsByCustomer = async (filter, pageSize = 50, offset = 0) => tatum_1.post(`/v3/ledger/transaction/customer?pageSize=${pageSize}&offset=${offset}`, filter, model_1.TransactionFilter);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/getTransactions" target="_blank">Tatum API documentation</a>
 */
exports.getTransactionsByLedger = async (filter, pageSize = 50, offset = 0) => tatum_1.post(`/v3/ledger/transaction/ledger?pageSize=${pageSize}&offset=${offset}`, filter, model_1.TransactionFilter);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/getTransactionsByAccountId" target="_blank">Tatum API documentation</a>
 */
exports.countTransactionsByAccount = async (filter) => tatum_1.post(`/v3/ledger/transaction/account?count=true`, filter, model_1.TransactionFilter);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/getTransactionsByCustomerId" target="_blank">Tatum API documentation</a>
 */
exports.countTransactionsByCustomer = async (filter) => tatum_1.post(`/v3/ledger/transaction/customer?count=true`, filter, model_1.TransactionFilter);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/getTransactions" target="_blank">Tatum API documentation</a>
 */
exports.countTransactionsByLedger = async (filter) => tatum_1.post(`/v3/ledger/transaction/ledger?count=true`, filter, model_1.TransactionFilter);
