// thorify/dist/provider/rpc-methods.js
'use strict';
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const thor_devkit_1 = require("thor-devkit");
const utils = require("../utils");
const simple_http_1 = require("./simple-http");
const debug = require('debug')('thor:http-provider:rpc');
exports.RPCMethodMap = new Map();
const getRevertReason = (returned) => {
    const ErrorStringMatcher = /^0x08c379a0/i;
    const PanicUint256Matcher = /^0x4e487b71/i;
    try {
        if (ErrorStringMatcher.test(returned)) {
            return thor_devkit_1.abi.decodeParameter('string', returned.replace(ErrorStringMatcher, '0x'));
        }
        if (PanicUint256Matcher.test(returned)) {
            return `Panic(0x${parseInt(thor_devkit_1.abi.decodeParameter('uint256', returned.replace(PanicUint256Matcher, '0x')), 10).toString(16)})`;
        }
        return null;
    }
    catch (_a) {
        return null;
    }
};
const HTTPPostProcessor = function (res) {
    if (res.Code === 0) {
        return Promise.reject(new Error(`[thor-provider] Invalid response, check the host`));
    }
    if (res.Code !== 200) {
        return Promise.reject(new Error(res.Body ? res.Body : ('[thor-provider] Invalid response code from provider: ' + res.Code)));
    }
    return Promise.resolve(res.Body);
};
exports.RPCMethodMap.set('eth_getBlockByNumber', function (rpc, host, timeout) {
    return __awaiter(this, void 0, void 0, function* () {
        const URL = host + '/blocks/' + utils.fromETHBlockNumber(rpc.params[0]);
        const res = yield simple_http_1.HTTP.get(URL, timeout).then(HTTPPostProcessor);
        return rpc.makeResult(res);
    });
});
exports.RPCMethodMap.set('eth_getBlockByHash', function (rpc, host, timeout) {
    return __awaiter(this, void 0, void 0, function* () {
        const URL = host + '/blocks/' + rpc.params[0];
        const res = yield simple_http_1.HTTP.get(URL, timeout).then(HTTPPostProcessor);
        return rpc.makeResult(res);
    });
});
exports.RPCMethodMap.set('eth_blockNumber', function (rpc, host, timeout) {
    return __awaiter(this, void 0, void 0, function* () {
        const URL = host + '/blocks/best';
        const res = yield simple_http_1.HTTP.get(URL, timeout).then(HTTPPostProcessor);
        return rpc.makeResult(!res ? null : res.number);
    });
});
exports.RPCMethodMap.set('eth_getBalance', function (rpc, host, timeout) {
    return __awaiter(this, void 0, void 0, function* () {
        const URL = host + '/accounts/' + rpc.params[0] + '?revision=' + utils.fromETHBlockNumberOrHash(rpc.params[1]);
        const res = yield simple_http_1.HTTP.get(URL, timeout).then(HTTPPostProcessor);
        return rpc.makeResult(!res ? null : res.balance);
    });
});
exports.RPCMethodMap.set('eth_getEnergy', function (rpc, host, timeout) {
    return __awaiter(this, void 0, void 0, function* () {
        const URL = host + '/accounts/' + rpc.params[0] + '?revision=' + utils.fromETHBlockNumberOrHash(rpc.params[1]);
        const res = yield simple_http_1.HTTP.get(URL, timeout).then(HTTPPostProcessor);
        return rpc.makeResult(!res ? null : res.energy);
    });
});
exports.RPCMethodMap.set('eth_getCode', function (rpc, host, timeout) {
    return __awaiter(this, void 0, void 0, function* () {
        const URL = host + '/accounts/' + rpc.params[0] + '/code?revision=' + utils.fromETHBlockNumberOrHash(rpc.params[1]);
        const res = yield simple_http_1.HTTP.get(URL, timeout).then(HTTPPostProcessor);
        return rpc.makeResult(!res ? null : res.code);
    });
});
exports.RPCMethodMap.set('eth_getStorageAt', function (rpc, host, timeout) {
    return __awaiter(this, void 0, void 0, function* () {
        const URL = host + '/accounts/' + rpc.params[0] + '/storage/' + utils.leftPadToBytes32(rpc.params[1]) + '?revision=' + utils.fromETHBlockNumberOrHash(rpc.params[2]);
        const res = yield simple_http_1.HTTP.get(URL, timeout).then(HTTPPostProcessor);
        return rpc.makeResult(!res ? null : res.value);
    });
});
exports.RPCMethodMap.set('eth_sendRawTransaction', function (rpc, host, timeout) {
    return __awaiter(this, void 0, void 0, function* () {
        const URL = host + '/transactions';
        const reqBody = {
            raw: rpc.params[0],
        };
        const res = yield simple_http_1.HTTP.post(URL, reqBody, timeout).then(HTTPPostProcessor);
        return rpc.makeResult(!res ? null : res.id);
    });
});
exports.RPCMethodMap.set('eth_getTransactionByHash', function (rpc, host, timeout) {
    return __awaiter(this, void 0, void 0, function* () {
        const URL = host + '/transactions/' + rpc.params[0];
        const res = yield simple_http_1.HTTP.get(URL, timeout).then(HTTPPostProcessor);
        if (!res) {
            return rpc.makeResult(null);
        }
        res.blockNumber = res.meta.blockNumber;
        return rpc.makeResult(res);
    });
});
exports.RPCMethodMap.set('eth_getTransactionReceipt', function (rpc, host, timeout) {
    return __awaiter(this, void 0, void 0, function* () {
        const URL = host + '/transactions/' + rpc.params[0] + '/receipt';
        const res = yield simple_http_1.HTTP.get(URL, timeout).then(HTTPPostProcessor);
        if (!res) {
            return rpc.makeResult(null);
        }
        res.blockNumber = res.meta.blockNumber;
        res.blockHash = res.meta.blockID;
        res.transactionHash = res.meta.txID;
        // For compatible with ethereum's receipt
        if (res.reverted) {
            res.status = '0x0';
        }
        else {
            res.status = '0x1';
        }
        if (res.outputs.length === 1) {
            res.contractAddress = res.outputs[0].contractAddress;
        }
        return rpc.makeResult(res);
    });
});
exports.RPCMethodMap.set('eth_call', function (rpc, host, timeout) {
    return __awaiter(this, void 0, void 0, function* () {
        const extraURI = '?revision=' + utils.fromETHBlockNumberOrHash(rpc.params[1]);
        const URL = host + '/accounts/*' + extraURI;
        const reqBody = {
            clauses: [{
                    to: rpc.params[0].to || null,
                    value: rpc.params[0].value || '',
                    data: rpc.params[0].data || '0x',
                }],
            gasPrice: rpc.params[0].gasPrice || undefined,
        };
        if (rpc.params[0].gas) {
            if (typeof rpc.params[0].gas === 'number') {
                reqBody.gas = rpc.params[0].gas;
            }
            else {
                reqBody.gas = parseInt(utils.sanitizeHex(rpc.params[0].gas), 16);
            }
        }
        if (rpc.params[0].from) {
            reqBody.caller = rpc.params[0].from;
        }
        const res = yield simple_http_1.HTTP.post(URL, reqBody, timeout).then(HTTPPostProcessor);
        debug('eth_call returns', res);
        if (!res || res.length === 0) {
            return rpc.makeResult(null);
        }
        else {
            const result = res[0];
            if (result.reverted || result.vmError) {
                if (result.data) {
                    const reason = getRevertReason(result.data);
                    if (reason) {
                        return rpc.makeError('VM reverted: ' + reason);
                    }
                }
                return rpc.makeError('VM executing failed' + (result.vmError ? ': ' + result.vmError : ''));
            }
            else {
                return rpc.makeResult(result.data === '0x' ? '' : result.data);
            }
        }
    });
});
exports.RPCMethodMap.set('eth_estimateGas', function (rpc, host, timeout) {
    return __awaiter(this, void 0, void 0, function* () {
        const extraURI = '?revision=' + utils.fromETHBlockNumberOrHash(rpc.params[1]);
        const URL = host + '/accounts/*' + extraURI;
        const reqBody = {
            clauses: [{
                    to: rpc.params[0].to || null,
                    value: rpc.params[0].value || '',
                    data: rpc.params[0].data || '0x',
                }],
            gasPrice: rpc.params[0].gasPrice || undefined,
        };
        if (rpc.params[0].gas) {
            if (typeof rpc.params[0].gas === 'number') {
                reqBody.gas = rpc.params[0].gas;
            }
            else {
                reqBody.gas = parseInt(utils.sanitizeHex(rpc.params[0].gas), 16);
            }
        }
        if (rpc.params[0].from) {
            reqBody.caller = rpc.params[0].from;
        }
        const res = yield simple_http_1.HTTP.post(URL, reqBody, timeout).then(HTTPPostProcessor);
        if (!res || res.length === 0) {
            return rpc.makeResult(null);
        }
        else {
            const result = res[0];
            if (result.reverted || result.vmError) {
                if (result.data) {
                    const reason = getRevertReason(result.data);
                    if (reason) {
                        return rpc.makeError('Gas estimation failed with VM reverted: ' + reason);
                    }
                }
                return rpc.makeError('Gas estimation failed' + (result.vmError ? ': ' + result.vmError : ''));
            }
            else {
                debug('VM gas:', result.gasUsed);
                // ignore the overflow since block gas limit is uint64 and JavaScript's max number is 2^53
                const intrinsicGas = thor_devkit_1.Transaction.intrinsicGas(reqBody.clauses);
                // increase vm gas by 15000 for safe since it's estimated from current block state, final state for the transaction is not determined for now
                return rpc.makeResult(intrinsicGas + (result.gasUsed ? (result.gasUsed + 15000) : 0));
            }
        }
    });
});
exports.RPCMethodMap.set('eth_getLogs', function (rpc, host, timeout) {
    return __awaiter(this, void 0, void 0, function* () {
        const reqBody = utils.formatLogQuery(rpc.params[0]);
        const URL = host + '/logs/event';
        const res = yield simple_http_1.HTTP.post(URL, reqBody, timeout).then(HTTPPostProcessor);
        if (!res) {
            return rpc.makeResult(null);
        }
        for (const item of res) {
            item.blockNumber = item.meta.blockNumber;
            item.blockHash = item.meta.blockID;
            item.transactionHash = item.meta.txID;
        }
        return rpc.makeResult(res);
    });
});
exports.RPCMethodMap.set('eth_getBlockRef', function (rpc, host, timeout) {
    return __awaiter(this, void 0, void 0, function* () {
        const URL = host + '/blocks/best';
        const res = yield simple_http_1.HTTP.get(URL, timeout).then(HTTPPostProcessor);
        if (!res || !res.id) {
            return rpc.makeResult(null);
        }
        return rpc.makeResult(res.id.substr(0, 18));
    });
});
exports.RPCMethodMap.set('eth_getChainTag', function (rpc, host, timeout) {
    return __awaiter(this, void 0, void 0, function* () {
        const URL = host + '/blocks/0';
        const res = yield simple_http_1.HTTP.get(URL, timeout).then(HTTPPostProcessor);
        if (!res || !res.id || res.id.length !== 66) {
            return rpc.makeResult(null);
        }
        return rpc.makeResult('0x' + res.id.substr(64, 2));
    });
});
