// @tatumio/tatum/dist/src/connector/tatum.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateBody = exports.httpDelete = exports.put = exports.post = exports.get = void 0;
const axios_1 = __importDefault(require("axios"));
const axios_retry_1 = __importStar(require("axios-retry"));
const class_transformer_1 = require("class-transformer");
const class_validator_1 = require("class-validator");
const constants_1 = require("../constants");
// In case of 429 Too Many Requests response error, request is triggered again
axios_retry_1.default(axios_1.default, {
    retryDelay: () => process.env.TATUM_RETRY_DELAY ? Number(process.env.TATUM_RETRY_DELAY) : constants_1.TATUM_RETRY_DELAY,
    retries: process.env.TATUM_RETRIES ? Number(process.env.TATUM_RETRIES) : constants_1.TATUM_RETRIES,
    retryCondition: (error) => { var _a; return axios_retry_1.isNetworkOrIdempotentRequestError(error) || ((_a = error === null || error === void 0 ? void 0 : error.response) === null || _a === void 0 ? void 0 : _a.status) === 429; }
});
const baseUrl = () => process.env.TATUM_API_URL || constants_1.TATUM_API_URL;
const headers = () => ({ headers: { 'x-api-key': process.env.TATUM_API_KEY } });
exports.get = async (url) => {
    const { data } = await axios_1.default.get(`${baseUrl()}${url}`, headers());
    return data;
};
exports.post = async (url, body, classType) => {
    await exports.validateBody(body, classType);
    const { data } = await axios_1.default.post(`${baseUrl()}${url}`, body, headers());
    return data;
};
exports.put = async (url, body, classType) => {
    await exports.validateBody(body, classType);
    const { data } = await axios_1.default.put(`${baseUrl()}${url}`, body, headers());
    return data;
};
exports.httpDelete = async (url) => {
    await axios_1.default.delete(`${baseUrl()}${url}`, headers());
};
exports.validateBody = async (body, classType) => {
    if (classType) {
        const classInstance = class_transformer_1.plainToClass(classType, body);
        await class_validator_1.validateOrReject(classInstance);
    }
};
