// @tatumio/tatum/dist/src/wallet/wallet.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateWallet = exports.generateAdaWallet = exports.generateLyraWallet = exports.generateXlmWallet = exports.generateXrpWallet = exports.generateNeoWallet = exports.generateLtcWallet = exports.generateTronWallet = exports.generateDogeWallet = exports.generateBtcWallet = exports.generateBchWallet = exports.generateCeloWallet = exports.generateBscWallet = exports.generateEthWallet = exports.generateVetWallet = exports.generateBnbWallet = void 0;
const crypto_1 = require("@binance-chain/javascript-sdk/lib/crypto");
const neon_js_1 = __importStar(require("@cityofzion/neon-js"));
const bip39_1 = require("bip39");
const bitcoinjs_lib_1 = require("bitcoinjs-lib");
const ethereumjs_wallet_1 = require("ethereumjs-wallet");
// @ts-ignore
const hdkey_1 = __importDefault(require("hdkey"));
const ripple_lib_1 = require("ripple-lib");
const stellar_sdk_1 = require("stellar-sdk");
const constants_1 = require("../constants");
const model_1 = require("../model");
const cardano_crypto_1 = __importDefault(require("./cardano.crypto"));
// tslint:disable-next-line:no-var-requires
const TronWeb = require('tronweb');
/**
 * Generate BnB wallet
 * @param testnet testnet or mainnet version of address
 * @returns wallet
 */
exports.generateBnbWallet = async (testnet) => {
    const privateKey = crypto_1.generatePrivateKey();
    const prefix = testnet ? 'tbnb' : 'bnb';
    return {
        address: crypto_1.getAddressFromPrivateKey(privateKey, prefix),
        privateKey,
    };
};
/**
 * Generate VeChain wallet
 * @param testnet testnet or mainnet version of address
 * @param mnem mnemonic seed to use
 * @returns wallet
 */
exports.generateVetWallet = async (testnet, mnem) => {
    const path = testnet ? constants_1.TESTNET_DERIVATION_PATH : constants_1.VET_DERIVATION_PATH;
    const hdwallet = ethereumjs_wallet_1.hdkey.fromMasterSeed(await bip39_1.mnemonicToSeed(mnem));
    const derivePath = hdwallet.derivePath(path);
    return {
        xpub: derivePath.publicExtendedKey().toString(),
        mnemonic: mnem
    };
};
/**
 * Generate Ethereum or any other ERC20 wallet
 * @param testnet testnet or mainnet version of address
 * @param mnem mnemonic seed to use
 * @returns wallet
 */
exports.generateEthWallet = async (testnet, mnem) => {
    const path = testnet ? constants_1.TESTNET_DERIVATION_PATH : constants_1.ETH_DERIVATION_PATH;
    const hdwallet = ethereumjs_wallet_1.hdkey.fromMasterSeed(await bip39_1.mnemonicToSeed(mnem));
    const derivePath = hdwallet.derivePath(path);
    return {
        xpub: derivePath.publicExtendedKey().toString(),
        mnemonic: mnem
    };
};
/**
 * Generate BSC or any other BEP-20 or BEP-721 wallet
 * @param testnet testnet or mainnet version of address
 * @param mnem mnemonic seed to use
 * @returns wallet
 */
exports.generateBscWallet = async (testnet, mnem) => {
    return exports.generateEthWallet(testnet, mnem);
};
/**
 * Generate Celo or any other ERC20 wallet
 * @param testnet testnet or mainnet version of address
 * @param mnem mnemonic seed to use
 * @returns wallet
 */
exports.generateCeloWallet = async (testnet, mnem) => {
    const path = testnet ? constants_1.TESTNET_DERIVATION_PATH : constants_1.CELO_DERIVATION_PATH;
    const hdwallet = ethereumjs_wallet_1.hdkey.fromMasterSeed(await bip39_1.mnemonicToSeed(mnem));
    const derivePath = hdwallet.derivePath(path);
    return {
        xpub: derivePath.publicExtendedKey().toString(),
        mnemonic: mnem
    };
};
/**
 * Generate Bitcoin Cash wallet
 * @param testnet testnet or mainnet version of address
 * @param mnem mnemonic seed to use
 * @returns wallet
 */
exports.generateBchWallet = async (testnet, mnem) => {
    const hdwallet = hdkey_1.default.fromMasterSeed(await bip39_1.mnemonicToSeed(mnem), testnet ? bitcoinjs_lib_1.networks.testnet.bip32 : bitcoinjs_lib_1.networks.bitcoin.bip32);
    return {
        mnemonic: mnem,
        xpub: hdwallet.derive(constants_1.BCH_DERIVATION_PATH).toJSON().xpub
    };
};
/**
 * Generate Bitcoin wallet
 * @param testnet testnet or mainnet version of address
 * @param mnem mnemonic seed to use
 * @returns wallet
 */
exports.generateBtcWallet = async (testnet, mnem) => {
    const hdwallet = hdkey_1.default.fromMasterSeed(await bip39_1.mnemonicToSeed(mnem), testnet ? bitcoinjs_lib_1.networks.testnet.bip32 : bitcoinjs_lib_1.networks.bitcoin.bip32);
    return {
        mnemonic: mnem,
        xpub: hdwallet.derive(testnet ? constants_1.TESTNET_DERIVATION_PATH : constants_1.BTC_DERIVATION_PATH).toJSON().xpub
    };
};
/**
 * Generate Doge wallet
 * @param testnet testnet or mainnet version of address
 * @param mnem mnemonic seed to use
 * @returns wallet
 */
exports.generateDogeWallet = async (testnet, mnem) => {
    const hdwallet = hdkey_1.default.fromMasterSeed(await bip39_1.mnemonicToSeed(mnem), testnet ? constants_1.DOGE_TEST_NETWORK.bip32 : constants_1.DOGE_NETWORK.bip32);
    return {
        mnemonic: mnem,
        xpub: hdwallet.derive(testnet ? constants_1.TESTNET_DERIVATION_PATH : constants_1.DOGE_DERIVATION_PATH).toJSON().xpub
    };
};
/**
 * Generate Tron wallet
 * @returns mnemonic for the wallet
 */
exports.generateTronWallet = async (mnem) => {
    const w = bitcoinjs_lib_1.bip32.fromSeed(await bip39_1.mnemonicToSeed(mnem));
    const bip32Interface = w.derivePath(constants_1.TRON_DERIVATION_PATH);
    return {
        mnemonic: mnem,
        xpub: bip32Interface.publicKey.toString('hex') + bip32Interface.chainCode.toString('hex')
    };
};
/**
 * Generate Litecoin wallet
 * @param testnet testnet or mainnet version of address
 * @param mnem mnemonic seed to use
 * @returns wallet
 */
exports.generateLtcWallet = async (testnet, mnem) => {
    const hdwallet = hdkey_1.default.fromMasterSeed(await bip39_1.mnemonicToSeed(mnem), testnet ? constants_1.LTC_TEST_NETWORK.bip32 : constants_1.LTC_NETWORK.bip32);
    return {
        mnemonic: mnem,
        xpub: hdwallet.derive(testnet ? constants_1.TESTNET_DERIVATION_PATH : constants_1.LTC_DERIVATION_PATH).toJSON().xpub
    };
};
/**
 * Generate Neo address and private key.
 */
exports.generateNeoWallet = () => {
    const privateKey = neon_js_1.default.create.privateKey();
    return { privateKey, address: new neon_js_1.wallet.Account(privateKey).address };
};
/**
 * Generate Xrp address and secret.
 */
exports.generateXrpWallet = () => {
    const { address, secret } = new ripple_lib_1.RippleAPI().generateAddress();
    return { address, secret };
};
/**
 * Generate Stellar address and secret.
 * @param secret secret of the account to generate address
 */
exports.generateXlmWallet = (secret) => {
    const keypair = secret ? stellar_sdk_1.Keypair.fromSecret(secret) : stellar_sdk_1.Keypair.random();
    return { address: keypair.publicKey(), secret: keypair.secret() };
};
/**
 * Generate Scrypta wallet
 * @param testnet testnet or mainnet version of address
 * @param mnem mnemonic seed to use
 * @returns wallet
 */
exports.generateLyraWallet = async (testnet, mnem) => {
    const hdwallet = hdkey_1.default.fromMasterSeed(await bip39_1.mnemonicToSeed(mnem), testnet ? constants_1.LYRA_TEST_NETWORK.bip32 : constants_1.LYRA_NETWORK.bip32);
    return { mnemonic: mnem, xpub: hdwallet.derive(constants_1.LYRA_DERIVATION_PATH).toJSON().xpub };
};
/**
 * Generate ADA wallet
 * @param mnemonic mnemonic seed to use
 * @returns wallet
 */
exports.generateAdaWallet = async (mnemonic) => {
    return { mnemonic, xpub: await cardano_crypto_1.default.generateXPublicKey(mnemonic) };
};
/**
 * Generate wallet
 * @param currency blockchain to generate wallet for
 * @param testnet testnet or mainnet version of address
 * @param mnemonic mnemonic seed to use. If not present, new one will be generated
 * @returns wallet or a combination of address and private key
 */
exports.generateWallet = (currency, testnet, mnemonic) => {
    const mnem = mnemonic ? mnemonic : bip39_1.generateMnemonic(256);
    switch (currency) {
        case model_1.Currency.BTC:
            return exports.generateBtcWallet(testnet, mnem);
        case model_1.Currency.DOGE:
            return exports.generateDogeWallet(testnet, mnem);
        case model_1.Currency.LTC:
            return exports.generateLtcWallet(testnet, mnem);
        case model_1.Currency.BCH:
            return exports.generateBchWallet(testnet, mnem);
        case model_1.Currency.TRON:
        case model_1.Currency.USDT_TRON:
            return exports.generateTronWallet(mnem);
        case model_1.Currency.CELO:
            return exports.generateCeloWallet(testnet, mnem);
        case model_1.Currency.USDT:
        case model_1.Currency.WBTC:
        case model_1.Currency.LEO:
        case model_1.Currency.LINK:
        case model_1.Currency.UNI:
        case model_1.Currency.FREE:
        case model_1.Currency.MKR:
        case model_1.Currency.USDC:
        case model_1.Currency.BAT:
        case model_1.Currency.TUSD:
        case model_1.Currency.PAX:
        case model_1.Currency.PAXG:
        case model_1.Currency.PLTC:
        case model_1.Currency.XCON:
        case model_1.Currency.ETH:
        case model_1.Currency.BSC:
        case model_1.Currency.BETH:
        case model_1.Currency.BBTC:
        case model_1.Currency.BADA:
        case model_1.Currency.WBNB:
        case model_1.Currency.BDOT:
        case model_1.Currency.BXRP:
        case model_1.Currency.BLTC:
        case model_1.Currency.BBCH:
        case model_1.Currency.MMY:
            return exports.generateEthWallet(testnet, mnem);
        case model_1.Currency.XRP:
            return exports.generateXrpWallet();
        case model_1.Currency.XLM:
            return exports.generateXlmWallet();
        case model_1.Currency.VET:
            return exports.generateVetWallet(testnet, mnem);
        case model_1.Currency.NEO:
            return exports.generateNeoWallet();
        case model_1.Currency.BNB:
            return exports.generateBnbWallet(testnet);
        case model_1.Currency.LYRA:
            return exports.generateLyraWallet(testnet, mnem);
        case model_1.Currency.ADA:
            return exports.generateAdaWallet(mnem);
        default:
            throw new Error('Unsupported blockchain.');
    }
};
