// @tatumio/tatum/dist/src/transaction/quorum.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.sendQuorumTransaction = exports.sendStoreDataQuorumTransaction = void 0;
const axios_1 = __importDefault(require("axios"));
const web3_utils_1 = require("web3-utils");
const tatum_1 = require("../connector/tatum");
const model_1 = require("../model");
const model_2 = require("../model");
/**
 * Send Quorum store data transaction to the blockchain.
 * This operation is irreversible.
 * @param body content of the transaction to broadcast
 * @param provider url of the Quorum Server to connect to.
 * @returns transaction id of the transaction in the blockchain
 */
exports.sendStoreDataQuorumTransaction = async (body, provider) => {
    await tatum_1.validateBody(body, model_1.CreateRecord);
    const data = (await axios_1.default.post(provider, {
        'jsonrpc': '2.0',
        'method': 'eth_sendTransaction',
        'params': [{
                from: body.from,
                to: body.to,
                nonce: body.nonce,
                data: body.data ? (web3_utils_1.isHex(body.data) ? web3_utils_1.stringToHex(body.data) : web3_utils_1.toHex(body.data)) : undefined,
            }],
        'id': 1
    })).data;
    if (data.result) {
        return { txId: data.result };
    }
    throw new Error(data.error.message);
};
/**
 * Send Quorum transaction to the blockchain.
 * This operation is irreversible.
 * @param body content of the transaction to broadcast
 * @param provider url of the Quorum Server to connect to.
 * @returns transaction id of the transaction in the blockchain
 */
exports.sendQuorumTransaction = async (body, provider) => {
    await tatum_1.validateBody(body, model_2.TransferQuorum);
    const data = (await axios_1.default.post(provider, {
        'jsonrpc': '2.0',
        'method': 'eth_sendTransaction',
        'params': [body],
        'id': 1
    })).data;
    if (data.result) {
        return { txId: data.result };
    }
    throw new Error(data.error.message);
};
