// @tatumio/tatum/dist/src/offchain/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.offchainCompleteWithdrawal = exports.offchainCancelWithdrawal = exports.offchainStoreWithdrawal = exports.offchainBroadcast = exports.getDepositAddressesForAccount = exports.removeDepositAddress = exports.registerTronTrc = exports.storeTrcContractAddress = exports.storeErc20ContractAddress = exports.registerEthereumErc20 = exports.assignDepositAddress = exports.getWithdrawals = exports.checkAddressExists = exports.generateDepositAddresses = exports.generateDepositAddress = void 0;
const tatum_1 = require("../connector/tatum");
const model_1 = require("../model");
const CreateOffchainAddressesBatch_1 = require("../model/request/CreateOffchainAddressesBatch");
const CreateTrcOffchain_1 = require("../model/request/CreateTrcOffchain");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/generateDepositAddress" target="_blank">Tatum API documentation</a>
 */
exports.generateDepositAddress = async (id, index) => {
    const url = `/v3/offchain/account/${id}/address`;
    return tatum_1.post(index === undefined ? url : `${url}?index=${index}`);
};
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/generateDepositAddressesBatch" target="_blank">Tatum API documentation</a>
 */
exports.generateDepositAddresses = async (batch) => tatum_1.post(`/v3/offchain/account/address/batch`, batch, CreateOffchainAddressesBatch_1.AddressBatch);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/addressExists" target="_blank">Tatum API documentation</a>
 */
exports.checkAddressExists = async (address, currency, index) => {
    const url = `/v3/offchain/account/address/${address}/${currency}`;
    return tatum_1.get(index === undefined ? url : `${url}?index=${index}`);
};
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/GetWithdrawals" target="_blank">Tatum API documentation</a>
 */
exports.getWithdrawals = async (status, currency, pageSize = 50, offset = 0) => {
    let url = `/v3/offchain/withdrawal?pageSize=${pageSize}&offset=${offset}`;
    if (status) {
        url += `&status=${status}`;
    }
    if (currency) {
        url += `&currency=${currency}`;
    }
    return tatum_1.get(url);
};
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/assignAddress" target="_blank">Tatum API documentation</a>
 */
exports.assignDepositAddress = async (id, address) => tatum_1.post(`/v3/offchain/account/${id}/address/${address}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/createErc20" target="_blank">Tatum API documentation</a>
 */
exports.registerEthereumErc20 = async (data) => tatum_1.post(`/v3/offchain/ethereum/erc20`, data, model_1.CreateErc20Offchain);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/storeErc20Address" target="_blank">Tatum API documentation</a>
 */
exports.storeErc20ContractAddress = async (name, address) => tatum_1.post(`/v3/offchain/ethereum/erc20/${name}/${address}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/storeTrcAddress" target="_blank">Tatum API documentation</a>
 */
exports.storeTrcContractAddress = async (name, address) => tatum_1.post(`/v3/offchain/tron/trc/${name}/${address}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/createTrc" target="_blank">Tatum API documentation</a>
 */
exports.registerTronTrc = async (data) => tatum_1.post(`/v3/offchain/tron/trc`, data, CreateTrcOffchain_1.CreateTrcOffchain);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/removeAddress" target="_blank">Tatum API documentation</a>
 */
exports.removeDepositAddress = async (id, address) => tatum_1.httpDelete(`/v3/offchain/account/${id}/address/${address}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/getAllDepositAddresses" target="_blank">Tatum API documentation</a>
 */
exports.getDepositAddressesForAccount = async (id) => tatum_1.get(`/v3/offchain/account/${id}/address`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/broadcastBlockchainTransaction" target="_blank">Tatum API documentation</a>
 */
exports.offchainBroadcast = async (data) => tatum_1.post(`/v3/offchain/withdrawal/broadcast`, data, model_1.BroadcastWithdrawal);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/storeWithdrawal" target="_blank">Tatum API documentation</a>
 */
exports.offchainStoreWithdrawal = async (data) => tatum_1.post(`/v3/offchain/withdrawal`, data);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/cancelInProgressWithdrawal" target="_blank">Tatum API documentation</a>
 */
exports.offchainCancelWithdrawal = async (id, revert = true) => tatum_1.httpDelete(`/v3/offchain/withdrawal/${id}?revert=${revert}`);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/completeWithdrawal" target="_blank">Tatum API documentation</a>
 */
exports.offchainCompleteWithdrawal = async (id, txId) => tatum_1.put(`/v3/offchain/withdrawal/${id}/${txId}`);
