// @tatumio/tatum/dist/src/offchain/kms.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.offchainTransferXrpKMS = exports.offchainTransferXlmKMS = exports.offchainTransferEthErc20KMS = exports.offchainTransferEthKMS = exports.offchainTransferBcashKMS = exports.offchainTransferLtcKMS = exports.offchainTransferBtcKMS = void 0;
const tatum_1 = require("../connector/tatum");
const TransferBtcBasedOffchainKMS_1 = require("../model/request/TransferBtcBasedOffchainKMS");
const TransferEthErc20OffchainKMS_1 = require("../model/request/TransferEthErc20OffchainKMS");
const TransferEthOffchainKMS_1 = require("../model/request/TransferEthOffchainKMS");
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/BtcTransfer" target="_blank">Tatum API documentation</a>
 */
exports.offchainTransferBtcKMS = async (body) => tatum_1.post(`/v3/offchain/bitcoin/transfer`, body, TransferBtcBasedOffchainKMS_1.TransferBtcBasedOffchainKMS);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/LtcTransfer" target="_blank">Tatum API documentation</a>
 */
exports.offchainTransferLtcKMS = async (body) => tatum_1.post(`/v3/offchain/litecoin/transfer`, body, TransferBtcBasedOffchainKMS_1.TransferBtcBasedOffchainKMS);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/BchTransfer" target="_blank">Tatum API documentation</a>
 */
exports.offchainTransferBcashKMS = async (body) => tatum_1.post(`/v3/offchain/bcash/transfer`, body, TransferBtcBasedOffchainKMS_1.TransferBtcBasedOffchainKMS);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/EthTransfer" target="_blank">Tatum API documentation</a>
 */
exports.offchainTransferEthKMS = async (body) => tatum_1.post(`/v3/offchain/ethereum/transfer`, body, TransferEthOffchainKMS_1.TransferEthOffchainKMS);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/EthTransferErc20" target="_blank">Tatum API documentation</a>
 */
exports.offchainTransferEthErc20KMS = async (body) => tatum_1.post(`/v3/offchain/ethereum/erc20/transfer`, body, TransferEthErc20OffchainKMS_1.TransferEthErc20OffchainKMS);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XlmTransfer" target="_blank">Tatum API documentation</a>
 */
exports.offchainTransferXlmKMS = async (body) => tatum_1.post(`/v3/offchain/xlm/transfer`, body, TransferBtcBasedOffchainKMS_1.TransferBtcBasedOffchainKMS);
/**
 * For more details, see <a href="https://tatum.io/apidoc#operation/XrpTransfer" target="_blank">Tatum API documentation</a>
 */
exports.offchainTransferXrpKMS = async (body) => tatum_1.post(`/v3/offchain/xrp/transfer`, body, TransferBtcBasedOffchainKMS_1.TransferBtcBasedOffchainKMS);
